File: annotationprocessor/src/main/java/org/acra/processor/element/AnnotationField.java
/*
 * Copyright (c) 2018 the ACRA team
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.acra.processor.element;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import com.squareup.javapoet.*;
import org.acra.processor.creator.BuildMethodCreator;
import org.acra.processor.util.ToCodeBlockVisitor;
import org.acra.processor.util.IsValidResourceVisitor;
import org.acra.processor.util.Strings;
import org.acra.processor.util.Types;
import org.apache.commons.text.WordUtils;

import javax.lang.model.element.AnnotationValue;
import javax.lang.model.element.Modifier;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collection;
import java.util.List;

/**
 * @author F43nd1r
 * @since 12.01.2018
 */

abstract class AnnotationField extends AbstractElement implements TransformedField.Transformable {
    private final Collection<ClassName> markers;
    private final String javadoc;
    private final AnnotationValue defaultValue;

    AnnotationField(@NonNull String name, @NonNull TypeName type, @NonNull Collection<AnnotationSpec> annotations, @Nullable String javadoc, @NonNull Collection<ClassName> markers, AnnotationValue defaultValue) {
        super(name, type, annotations);
        this.javadoc = javadoc;
        this.markers = markers;
        this.defaultValue = defaultValue;
    }

    boolean hasMarker(@NonNull ClassName marker) {
        return markers.contains(marker);
    }

    @Override
    public final void addToBuilder(@NonNull TypeSpec.Builder builder, @NonNull ClassName builderName, @NonNull CodeBlock.Builder constructorAlways, @NonNull CodeBlock.Builder constructorWhenAnnotationPresent, CodeBlock.Builder constructorWhenAnnotationMissing) {
        addWithoutGetter(builder, builderName, constructorAlways, constructorWhenAnnotationPresent, constructorWhenAnnotationMissing);
        addGetter(builder);
    }

    @Override
    public final void addWithoutGetter(@NonNull TypeSpec.Builder builder, ClassName builderName, CodeBlock.Builder constructorAlways, CodeBlock.Builder constructorWhenAnnotationPresent, CodeBlock.Builder constructorWhenAnnotationMissing) {
        TransformedField.Transformable.super.addToBuilder(builder, builderName, constructorAlways, constructorWhenAnnotationPresent, constructorWhenAnnotationMissing);
        addSetter(builder, builderName);
        addInitializer(constructorWhenAnnotationPresent, constructorWhenAnnotationMissing);
    }

    protected abstract void addInitializer(CodeBlock.Builder constructorWhenAnnotationPresent, CodeBlock.Builder constructorWhenAnnotationMissing);

    AnnotationValue getDefaultValue() {
        return defaultValue;
    }

    @Override
    public void configureSetter(@NonNull MethodSpec.Builder builder) {
        if (javadoc != null) {
            final String name = builder.build().parameters.get(0).name;
            builder.addJavadoc(javadoc.replaceAll("(\n|^) ", "$1").replaceAll("@return ((.|\n)*)$", "@param " + name + " $1@return this instance\n"));
        }
    }

    static class Normal extends AnnotationField {

        Normal(String name, TypeName type, Collection<AnnotationSpec> annotations, Collection<ClassName> markers, AnnotationValue defaultValue, String javadoc) {
            super(name, type, annotations, javadoc, markers, defaultValue);
        }

        @Override
        public void addInitializer(@NonNull CodeBlock.Builder constructorWhenAnnotationPresent, CodeBlock.Builder constructorWhenAnnotationMissing) {
            constructorWhenAnnotationPresent.addStatement("$1L = $2L.$1L()", getName(), Strings.VAR_ANNOTATION);
            if (getDefaultValue() != null) {
                constructorWhenAnnotationMissing.addStatement("$L = $L", getName(), getDefaultValue().accept(new ToCodeBlockVisitor(getType()), null));
            }
        }

        @Override
        public void addToBuildMethod(@NonNull BuildMethodCreator method) {
            if (getDefaultValue() == null) {
                method.addNotUnset(getName(), getType());
            }
            if (hasMarker(Types.NON_EMPTY)) {
                method.addNotEmpty(getName());
            }
            if (hasMarker(Types.INSTANTIATABLE)) {
                method.addInstantiatable(getName());
            }
            if (hasMarker(Types.ANY_NON_DEFAULT)) {
                method.addAnyNonDefault(getName(), getDefaultValue().accept(new ToCodeBlockVisitor(getType()), null));
            }
        }

    }

    static class StringResource extends AnnotationField {
        private final String originalName;
        private final boolean hasDefault;

        StringResource(String name, Collection<AnnotationSpec> annotations, Collection<ClassName> markers,
                       AnnotationValue defaultValue, String javadoc) {
            super((name.startsWith(Strings.PREFIX_RES)) ? WordUtils.uncapitalize(name.substring(Strings.PREFIX_RES.length())) : name, Types.STRING, annotations, javadoc, markers, defaultValue);
            this.originalName = name;
            this.hasDefault = defaultValue != null && getDefaultValue().accept(new IsValidResourceVisitor(), null);
            getAnnotations().remove(Types.STRING_RES);
            getAnnotations().add(hasDefault ? Types.NON_NULL : Types.NULLABLE);
        }

        @Override
        public void addInitializer(@NonNull CodeBlock.Builder constructorWhenAnnotationPresent, CodeBlock.Builder constructorWhenAnnotationMissing) {
            constructorWhenAnnotationPresent.beginControlFlow("if ($L.$L() != 0)", Strings.VAR_ANNOTATION, originalName)
                    .addStatement("$L = $L.getString($L.$L())", getName(), Strings.FIELD_CONTEXT, Strings.VAR_ANNOTATION, originalName)
                    .endControlFlow();
            if (hasDefault) {
                constructorWhenAnnotationMissing.addStatement("$L = $L.getString($L)", getName(), Strings.FIELD_CONTEXT, getDefaultValue());
            }

        }

        @Override
        public void addSetter(@NonNull TypeSpec.Builder builder, @NonNull ClassName builderName) {
            super.addSetter(builder, builderName);
            final MethodSpec.Builder setter = baseResSetter(builderName)
                    .addStatement("this.$L = $L.getString($L)", getName(), Strings.FIELD_CONTEXT, Strings.PREFIX_RES + WordUtils.capitalize(getName()))
                    .addStatement("return this");
            configureSetter(setter);
            builder.addMethod(setter.build());
        }

        private MethodSpec.Builder baseResSetter(ClassName builderName) {
            final String parameterName = Strings.PREFIX_RES + WordUtils.capitalize(getName());
            final List<AnnotationSpec> annotations = new ArrayList<>(getAnnotations());
            annotations.removeIf(Arrays.asList(Types.NULLABLE, Types.NON_NULL)::contains);
            annotations.add(Types.STRING_RES);
            return MethodSpec.methodBuilder(Strings.PREFIX_SETTER + WordUtils.capitalize(parameterName))
                    .addParameter(ParameterSpec.builder(TypeName.INT, parameterName).addAnnotations(annotations).build())
                    .addModifiers(Modifier.PUBLIC)
                    .addAnnotation(Types.NON_NULL)
                    .returns(builderName);
        }

        @Override
        public void addToBuilderInterface(@NonNull TypeSpec.Builder builder, @NonNull ClassName builderName) {
            super.addToBuilderInterface(builder, builderName);
            final MethodSpec.Builder setter = baseResSetter(builderName).addModifiers(Modifier.ABSTRACT);
            configureSetter(setter);
            builder.addMethod(setter.build());
        }

        @Override
        public void addToBuildMethod(@NonNull BuildMethodCreator method) {
            if (getDefaultValue() == null) {
                method.addNotUnset(getName(), getType());
            }
            if (hasMarker(Types.ANY_NON_DEFAULT)) {
                method.addAnyNonDefault(getName(), CodeBlock.of("null"));
            }
        }
    }

}

File: annotationprocessor/src/main/java/org/acra/processor/element/AnnotationField.kt
/*
 * Copyright (c) 2018 the ACRA team
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.acra.processor.element

import com.squareup.kotlinpoet.AnnotationSpec
import com.squareup.kotlinpoet.ClassName
import com.squareup.kotlinpoet.CodeBlock
import com.squareup.kotlinpoet.FunSpec
import com.squareup.kotlinpoet.KModifier
import com.squareup.kotlinpoet.ParameterSpec
import com.squareup.kotlinpoet.ParameterizedTypeName
import com.squareup.kotlinpoet.ParameterizedTypeName.Companion.parameterizedBy
import com.squareup.kotlinpoet.PropertySpec
import com.squareup.kotlinpoet.TypeName
import com.squareup.kotlinpoet.TypeSpec
import com.squareup.kotlinpoet.WildcardTypeName
import org.acra.processor.creator.BuildMethodCreator
import org.acra.processor.util.IsValidResourceVisitor
import org.acra.processor.util.Strings
import org.acra.processor.util.ToCodeBlockVisitor
import org.acra.processor.util.Types
import org.apache.commons.text.WordUtils
import java.util.*
import javax.lang.model.element.AnnotationValue

/**
 * @author F43nd1r
 * @since 12.01.2018
 */
abstract class AnnotationField(override val name: String, override val type: TypeName, override val annotations: Collection<AnnotationSpec>, private val javadoc: String?,
                               private val markers: Collection<ClassName>, val defaultValue: AnnotationValue?) : TransformedField.Transformable {
    fun hasMarker(marker: ClassName): Boolean {
        return markers.contains(marker)
    }

    override fun addToBuilder(builder: TypeSpec.Builder, builderName: ClassName, constructor: CodeBlock.Builder) {
        super.addToBuilder(builder, builderName, constructor)
        addSetter(builder, builderName)
        addInitializer(constructor)
    }

    override fun configureField(builder: PropertySpec.Builder) {
        if (javadoc != null) {
            builder.addKdoc("%L", javadoc.replace("(\n|^) ".toRegex(), "$1"))
        }
    }

    protected abstract fun addInitializer(constructor: CodeBlock.Builder)
    override fun configureSetter(builder: FunSpec.Builder) {
        if (javadoc != null) {
            val name = builder.build().parameters[0].name
            builder.addKdoc("%L", javadoc.replace("(\n|^) ".toRegex(), "$1").replace("@return ((.|\n)*)$".toRegex(), "@param $name $1@return this instance\n"))
        }
    }

    open class Normal(name: String, type: TypeName, annotations: Collection<AnnotationSpec>, markers: Collection<ClassName>, defaultValue: AnnotationValue?, javadoc: String) :
            AnnotationField(name, type, annotations, javadoc, markers, defaultValue) {
        public override fun addInitializer(constructor: CodeBlock.Builder) {
            constructor.addStatement("%1L = %2L?.%1L ?: %3L", name, Strings.VAR_ANNOTATION, defaultValue?.accept(ToCodeBlockVisitor(), null))
        }

        override fun addToBuildMethod(method: BuildMethodCreator) {
            if (defaultValue == null) {
                method.addNotUnset(name)
            }
            if (hasMarker(Types.NON_EMPTY)) {
                method.addNotEmpty(name)
            }
            if (hasMarker(Types.INSTANTIATABLE)) {
                method.addInstantiatable(name)
            }
            if (hasMarker(Types.ANY_NON_DEFAULT)) {
                method.addAnyNonDefault(name, defaultValue?.accept(ToCodeBlockVisitor(), null) ?: CodeBlock.of("null"))
            }
        }
    }

    class Clazz(name: String, type: TypeName, annotations: Collection<AnnotationSpec>, markers: Collection<ClassName>, defaultValue: AnnotationValue?, javadoc: String) :
            Normal(name, type, annotations, markers, defaultValue, javadoc) {
        override fun addInitializer(constructor: CodeBlock.Builder) {
            constructor.addStatement("%1L = %2L?.%1L?.java ?: %3L", name, Strings.VAR_ANNOTATION, defaultValue?.accept(ToCodeBlockVisitor(), null))
        }
    }

    class Array(name: String, type: TypeName, annotations: Collection<AnnotationSpec>, markers: Collection<ClassName>, defaultValue: AnnotationValue?, javadoc: String) :
            Normal(name, (type as ParameterizedTypeName).let { it.rawType.parameterizedBy(WildcardTypeName.producerOf(it.typeArguments.first())) }, annotations, markers,
                    defaultValue, javadoc) {
        private val originalType = type
        override fun configureSetter(builder: FunSpec.Builder) {
            super.configureSetter(builder)
            val param = builder.parameters.first().toBuilder(type = (originalType as ParameterizedTypeName).typeArguments.first()).addModifiers(KModifier.VARARG).build()
            builder.parameters.clear()
            builder.addParameter(param)
        }
    }

    class StringResource(private val originalName: String, annotations: Collection<AnnotationSpec>, markers: Collection<ClassName>,
                         defaultValue: AnnotationValue?, javadoc: String?) :
            AnnotationField(if (originalName.startsWith(Strings.PREFIX_RES)) WordUtils.uncapitalize(originalName.substring(Strings.PREFIX_RES.length)) else originalName,
                    Types.STRING, annotations - Types.STRING_RES, javadoc, markers, defaultValue) {
        private val hasDefault: Boolean = defaultValue != null && defaultValue.accept(IsValidResourceVisitor(), null)

        public override fun addInitializer(constructor: CodeBlock.Builder) {
            if (hasDefault) {
                constructor.addStatement("%L = %L.getString(%L?.%L.takeIf·{ it != 0 } ?: %L)", name, Strings.FIELD_CONTEXT, Strings.VAR_ANNOTATION, originalName, defaultValue)
            } else {
                constructor.addStatement("%L = %L?.%L.takeIf·{ it != 0 }?.let·{ %L.getString(it) } ?: \"\"", name, Strings.VAR_ANNOTATION, originalName, Strings.FIELD_CONTEXT)
            }
        }

        override fun addSetter(builder: TypeSpec.Builder, builderName: ClassName) {
            super.addSetter(builder, builderName)
            val setter = baseResSetter(builderName)
                    .addStatement("this.%L = %L.getString(%L)", name, Strings.FIELD_CONTEXT, Strings.ensurePrefix(Strings.PREFIX_RES, name))
                    .addStatement("return this")
            configureSetter(setter)
            builder.addFunction(setter.build())
        }

        private fun baseResSetter(builderName: ClassName): FunSpec.Builder {
            val parameterName = Strings.ensurePrefix(Strings.PREFIX_RES, name)
            val annotations: MutableList<AnnotationSpec> = ArrayList(annotations)
            annotations.add(Types.STRING_RES)
            return FunSpec.builder(Strings.ensurePrefix(Strings.PREFIX_SETTER, parameterName))
                    .addParameter(ParameterSpec.builder(parameterName, Int::class).addAnnotations(annotations).build())
                    .addModifiers(KModifier.PUBLIC)
                    .returns(builderName)
        }

        override fun addToBuildMethod(method: BuildMethodCreator) {
            if (defaultValue == null) {
                method.addNotUnset(name)
            }
            if (hasMarker(Types.ANY_NON_DEFAULT)) {
                method.addAnyNonDefault(name, CodeBlock.of("\"\""))
            }
        }
    }
}